pkg/channel/consolidated/status/status.go

```diff
@@ -182,44 +186,25 @@ func (m *Prober) IsReady(ctx context.Context, ch messagingv1beta1.KafkaChannel,
 			zap.Any("subscription", sub.UID))
 		return false, err
 	}
-
-	if ready, ok := func() (bool, bool) {
+	state, ok := func() (*targetState, bool) {
 		m.mu.Lock()
 		defer m.mu.Unlock()
-		if state, ok := m.targetStates[subscriptionKey]; ok {
-			if state.sub.Generation == sub.Generation {
-				state.lastAccessed = time.Now()
-				logger.Debugw("Subscription is cached. Checking readiness",
-					zap.Any("subscription", sub.UID))
-				if state.ready {
-					/* This subscription became ready once, that's sufficient for us for now.
-					Ideally we'd have a way to know if each partition already have at least one offset commit, but that's
-					another problem for another day
-					*/
-					return true, true
-				} else {
-					if target.PodIPs.Len() == state.probedPods.Len() && target.PodIPs.HasAll(state.probedPods.List()...) {
-						return false, true
-					} else {
-						logger.Debugw("Dispatcher pods changed. Canceling old probes and restarting",
-							zap.Any("subscription", sub.UID))
-						// the dispatcher pods changed, we need to cancel old probing and start again
-						m.mu.Unlock()
-						m.CancelProbing(state.sub)
-						m.mu.Lock()
-						return false, false
-					}
-				}
-			}
-			// Cancel the polling for the outdated version
-			state.cancel()
-			delete(m.targetStates, subscriptionKey)
+		s, o := m.targetStates[sub.UID]
+		return s, o
+	}()
```

Comment:
because of the lock, accessing the map needs to happen under lock and this construct of lock, defer ulock, inside an anonymous function is both idiomatic and less bug prone (since unlocking is deferred)

Reply:
the whole idea is to rely on go's defer reliability which the go runtime will ensure to take place even if the call panics, something that the normal call to 'unlock' doesn't guarantee. Check the example by the end here https://blog.golang.org/defer-panic-and-recover

Reply:
perhaps adding a comment, on why?

the debate here shows, there is reasons, but also questions.  Comments usually help to get larger context

Reply:
sure, will add a comment